--- Mesa-7.11/configure.ac~	2011-07-12 12:57:43.000000000 +0200
+++ Mesa-7.11/configure.ac	2011-07-18 19:40:58.701272226 +0200
@@ -908,7 +908,7 @@
 AC_SUBST([GLESv2_LIB_DEPS])
 AC_SUBST([GLESv2_PC_LIB_PRIV])
 
-GLAPI_LIB_DEPS="-lpthread"
+GLAPI_LIB_DEPS="-lpthread $SELINUX_LIBS"
 AC_SUBST([GLAPI_LIB_DEPS])
 
 
--- Mesa-7.11/configure.ac~	2011-08-01 13:49:24.000000000 +0200
+++ Mesa-7.11/configure.ac	2011-08-01 13:50:55.554153435 +0200
@@ -893,7 +893,7 @@
     fi
 
     # need DRM libs, -lpthread, etc.
-    GL_LIB_DEPS="$GL_LIB_DEPS $LIBDRM_LIBS -lm -lpthread $DLOPEN_LIBS"
+    GL_LIB_DEPS="$GL_LIB_DEPS $LIBDRM_LIBS -lm -lpthread $DLOPEN_LIBS $SELINUX_LIBS"
     GL_PC_LIB_PRIV="-lm -lpthread $DLOPEN_LIBS"
     GLESv1_CM_LIB_DEPS="$LIBDRM_LIBS -lm -lpthread $DLOPEN_LIBS"
     GLESv1_CM_PC_LIB_PRIV="-lm -lpthread $DLOPEN_LIBS"
